File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/changes/LectureChangesArrayAdapter.java
package nerd.tuxmobil.fahrplan.congress.changes;

import android.content.Context;
import android.graphics.Paint;
import android.support.annotation.ColorInt;
import android.support.v4.content.ContextCompat;
import android.view.View;
import android.widget.TextView;

import java.util.List;

import nerd.tuxmobil.fahrplan.congress.R;
import nerd.tuxmobil.fahrplan.congress.base.LecturesAdapter;
import nerd.tuxmobil.fahrplan.congress.models.Lecture;
import nerd.tuxmobil.fahrplan.congress.utils.DateHelper;

public class LectureChangesArrayAdapter extends LecturesAdapter {

    @ColorInt
    private int scheduleChangeTextColor;

    @ColorInt
    private int scheduleChangeNewTextColor;

    @ColorInt
    private int scheduleChangeCanceledTextColor;

    LectureChangesArrayAdapter(Context context, List<Lecture> list, int numDays) {
        super(context, R.layout.lecture_change_row, list, numDays);
        scheduleChangeTextColor = ContextCompat.getColor(context, R.color.schedule_change);
        scheduleChangeNewTextColor = ContextCompat.getColor(context, R.color.schedule_change_new);
        scheduleChangeCanceledTextColor = ContextCompat.getColor(context, R.color.schedule_change_canceled);
    }

    @Override
    protected void resetTextStyle(TextView textView, int style) {
        super.resetTextStyle(textView, style);
        textView.setPaintFlags(textView.getPaintFlags() & ~Paint.STRIKE_THRU_TEXT_FLAG);
    }

    private void setTextStyleChanged(TextView textView) {
        textView.setTextColor(scheduleChangeTextColor);
    }

    private void setTextStyleNew(TextView textView) {
        textView.setTextColor(scheduleChangeNewTextColor);
    }

    private void setTextStyleCanceled(TextView textView) {
        textView.setTextColor(scheduleChangeCanceledTextColor);
        textView.setPaintFlags(textView.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
    }

    @Override
    protected void initViewSetup() {
        // Nothing to do here
    }

    @Override
    protected void setItemContent(int position, ViewHolder viewHolder) {
        resetItemStyles(viewHolder);

        Lecture l = getLecture(position);
        viewHolder.title.setText(l.title);
        viewHolder.subtitle.setText(l.subtitle);
        viewHolder.speakers.setText(l.getFormattedSpeakers());
        viewHolder.lang.setText(l.lang);
        String dayText = DateHelper.getFormattedDate(l.dateUTC);
        viewHolder.day.setText(dayText);
        String timeText = DateHelper.getFormattedTime(l.dateUTC);
        viewHolder.time.setText(timeText);
        viewHolder.room.setText(l.room);
        String durationText = context.getString(R.string.event_duration, l.duration);
        viewHolder.duration.setText(durationText);
        viewHolder.video.setVisibility(View.GONE);
        viewHolder.noVideo.setVisibility(View.GONE);

        if (l.changedIsNew) {
            setTextStyleNew(viewHolder.title);
            setTextStyleNew(viewHolder.subtitle);
            setTextStyleNew(viewHolder.speakers);
            setTextStyleNew(viewHolder.lang);
            setTextStyleNew(viewHolder.day);
            setTextStyleNew(viewHolder.time);
            setTextStyleNew(viewHolder.room);
            setTextStyleNew(viewHolder.duration);
        } else if (l.changedIsCanceled) {
            setTextStyleCanceled(viewHolder.title);
            setTextStyleCanceled(viewHolder.subtitle);
            setTextStyleCanceled(viewHolder.speakers);
            setTextStyleCanceled(viewHolder.lang);
            setTextStyleCanceled(viewHolder.day);
            setTextStyleCanceled(viewHolder.time);
            setTextStyleCanceled(viewHolder.room);
            setTextStyleCanceled(viewHolder.duration);
        } else {
            if (l.changedTitle) {
                setTextStyleChanged(viewHolder.title);
                if (l.title.length() == 0) {
                    viewHolder.title.setText(context.getText(R.string.dash));
                }
            }
            if (l.changedSubtitle) {
                setTextStyleChanged(viewHolder.subtitle);
                if (l.subtitle.length() == 0) {
                    viewHolder.subtitle.setText(context.getText(R.string.dash));
                }
            }
            if (l.changedSpeakers) {
                setTextStyleChanged(viewHolder.speakers);
                if (l.speakers.length() == 0) {
                    viewHolder.speakers.setText(context.getText(R.string.dash));
                }
            }
            if (l.changedLanguage) {
                setTextStyleChanged(viewHolder.lang);
                if (l.lang.length() == 0) {
                    viewHolder.lang.setText(context.getText(R.string.dash));
                }
            }
            if (l.changedDay) {
                setTextStyleChanged(viewHolder.day);
            }
            if (l.changedTime) {
                setTextStyleChanged(viewHolder.time);
            }
            if (l.changedRoom) {
                setTextStyleChanged(viewHolder.room);
            }
            if (l.changedDuration) {
                setTextStyleChanged(viewHolder.duration);
            }
            if (l.changedRecordingOptOut) {
                if (l.recordingOptOut) {
                    viewHolder.noVideo.setVisibility(View.VISIBLE);
                } else {
                    viewHolder.video.setVisibility(View.VISIBLE);
                }
            }
        }
    }

}

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/changes/LectureChangesArrayAdapter.kt
package nerd.tuxmobil.fahrplan.congress.changes

import android.content.Context
import android.graphics.Paint
import android.view.View
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import info.metadude.android.eventfahrplan.commons.temporal.DateFormatter
import nerd.tuxmobil.fahrplan.congress.R
import nerd.tuxmobil.fahrplan.congress.base.LecturesAdapter
import nerd.tuxmobil.fahrplan.congress.extensions.textOrHide
import nerd.tuxmobil.fahrplan.congress.models.Lecture

class LectureChangesArrayAdapter internal constructor(

        context: Context,
        list: List<Lecture>,
        numDays: Int

) : LecturesAdapter(

        context,
        R.layout.lecture_list_item,
        list,
        numDays

) {

    @ColorInt
    private val scheduleChangeTextColor = ContextCompat.getColor(context, R.color.schedule_change)

    @ColorInt
    private val scheduleChangeNewTextColor = ContextCompat.getColor(context, R.color.schedule_change_new)

    @ColorInt
    private val scheduleChangeCanceledTextColor = ContextCompat.getColor(context, R.color.schedule_change_canceled)

    override fun resetTextStyle(textView: TextView, style: Int) {
        super.resetTextStyle(textView, style)
        textView.paintFlags = textView.paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
    }

    override fun initViewSetup() {
        // Nothing to do here
    }

    override fun setItemContent(position: Int, viewHolder: ViewHolder) {
        resetItemStyles(viewHolder)

        val lecture = getLecture(position)
        with(viewHolder) {
            title.textOrHide = lecture.title
            subtitle.textOrHide = lecture.subtitle
            speakers.textOrHide = lecture.formattedSpeakers
            lang.textOrHide = lecture.lang
            lang.contentDescription = lecture.getLanguageContentDescription(context)
            val dayText = DateFormatter.newInstance().getFormattedDate(lecture.dateUTC)
            day.textOrHide = dayText
            val timeText = DateFormatter.newInstance().getFormattedTime(lecture.dateUTC)
            time.textOrHide = timeText
            room.textOrHide = lecture.room
            val durationText = context.getString(R.string.event_duration, lecture.duration)
            duration.textOrHide = durationText
            video.visibility = View.GONE
            noVideo.visibility = View.GONE
            withoutVideoRecording.visibility = View.GONE

            if (lecture.changedIsNew) {
                title.setTextStyleNew()
                subtitle.setTextStyleNew()
                speakers.setTextStyleNew()
                lang.setTextStyleNew()
                day.setTextStyleNew()
                time.setTextStyleNew()
                room.setTextStyleNew()
                duration.setTextStyleNew()
            } else if (lecture.changedIsCanceled) {
                title.setTextStyleCanceled()
                subtitle.setTextStyleCanceled()
                speakers.setTextStyleCanceled()
                lang.setTextStyleCanceled()
                day.setTextStyleCanceled()
                time.setTextStyleCanceled()
                room.setTextStyleCanceled()
                duration.setTextStyleCanceled()
            } else {
                if (lecture.changedTitle) {
                    title.setTextStyleChanged()
                    if (lecture.title.isEmpty()) {
                        title.text = context.getText(R.string.dash)
                    }
                }
                if (lecture.changedSubtitle) {
                    subtitle.setTextStyleChanged()
                    if (lecture.subtitle.isEmpty()) {
                        subtitle.text = context.getText(R.string.dash)
                    }
                }
                if (lecture.changedSpeakers) {
                    speakers.setTextStyleChanged()
                    if (lecture.speakers.isEmpty()) {
                        speakers.text = context.getText(R.string.dash)
                    }
                }
                if (lecture.changedLanguage) {
                    lang.setTextStyleChanged()
                    if (lecture.lang.isEmpty()) {
                        lang.text = context.getText(R.string.dash)
                        lang.contentDescription = context.getText(R.string.lecture_list_item_language_removed_content_description)
                    }
                }
                if (lecture.changedDay) {
                    day.setTextStyleChanged()
                }
                if (lecture.changedTime) {
                    time.setTextStyleChanged()
                }
                if (lecture.changedRoom) {
                    room.setTextStyleChanged()
                }
                if (lecture.changedDuration) {
                    duration.setTextStyleChanged()
                }
                if (lecture.changedRecordingOptOut) {
                    if (lecture.recordingOptOut) {
                        noVideo.visibility = View.VISIBLE
                    } else {
                        video.visibility = View.VISIBLE
                    }
                }
            }
        }
    }

    private fun TextView.setTextStyleNew() = setTextColor(scheduleChangeNewTextColor)

    private fun TextView.setTextStyleCanceled() {
        setTextColor(scheduleChangeCanceledTextColor)
        paintFlags = paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
    }

    private fun TextView.setTextStyleChanged() = setTextColor(scheduleChangeTextColor)

}
